[TRTLLM-7349][feat] Adding new orchestrator type --  ray

joyang-nv opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Optional Ray-based orchestration for inference (multi-node, disaggregated, async).
  - Non-MPI path using PyTorch ProcessGroup, DeviceMesh-based mapping, and PG-backed collectives (allgather/allreduce/reducescatter).
  - Ray executor and GPU worker with queue-based result delivery.
  - LLM API: collective RPC and async/sync weight update helpers.
- Documentation
  - New READMEs for Ray workflows and disaggregated serving.
- Examples
  - Ray inference scripts, multi-node cluster launcher, disagg-serving CLI.
- Tests
  - Ray-enabled unit/integration tests and pytest markers/flag.
- Chores
  - Added ray dependency; build/packaging updates for new utilities.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

joyang-nv commented on requirements.txt:
```diff
@@ -68,3 +68,4 @@ triton==3.3.1; platform_machine == "x86_64"
 tiktoken
 blobfile
 openai-harmony==0.0.4
+ray[default]
```
Sorry, I mean no specific version is fine.

joyang-nv commented:
/bot run

tensorrt-cicd commented:
[PR_Github #17676](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17676/) [ run ] triggered by Bot

hchings commented:
/bot run --stage-list "H100_PCIe-PyTorch-1, H100_PCIe-PyTorch-Ray-1" --disable-fail-fast

joyang-nv commented:
/bot run

litaotju commented:
/bot run

joyang-nv commented:
/bot run

tensorrt-cicd commented:
[PR_Github #17989](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17989/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #17989](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17989/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13483](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13483/) completed with status: 'FAILURE'

joyang-nv commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #18019](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18019/) [ run ] triggered by Bot

tongyuantongyu commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #18172](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18172/) [ run ] completed with state `ABORTED`

tongyuantongyu commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #18176](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18176/) [ run ] completed with state `ABORTED`

tongyuantongyu commented:
/bot run --disable-fail-fast

lucaslie reviewed (approved):
approving on behalf of @NVIDIA/trt-llm-torch-autodeploy-devs  for the auto-deploy changes. didn't check the rest though. Please get separate reviews for the groups. I see my approval is also on behalf of a bunch of other groups

hchings commented on examples/ray/README.md:
```diff
@@ -0,0 +1,43 @@
+
+
+## Disclaimer
+The code is experimental and subject to change. Currently, there are no guarantees regarding functionality, performance, or stability.
```
updated

hchings commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #18500](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18500/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18500](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18500/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #13885](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13885/) completed with status: 'FAILURE'

hchings commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #18502](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18502/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18502](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18502/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13887](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13887/) completed with status: 'FAILURE'

hchings commented:
/bot run --stage-list "GB200-4_GPUs-PyTorch-1" --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #18518](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18518/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18519](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18519/) [ kill ] triggered by Bot

hchings commented:
/bot run --stage-list "GB200-4_GPUs-PyTorch-1" --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #18520](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18520/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18520](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18520/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13902 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13902/) completed with status: 'SUCCESS'

hchings commented on tests/integration/defs/llmapi/ray/test_llm_ray_e2e.py:
I'll remove this file. Eventually we use the existing torch backend e2e tests (`defs/llmapi/test_llm_examples.py`) for testing.

chuangz0 commented on examples/ray/disaggregated/disagg_serving_local.sh:
```diff
@@ -0,0 +1,141 @@
+  kv_cache_config:
+    free_gpu_memory_fraction: 0.2
+  cache_transceiver_config:
+    backend: "ucx"
```
should be `backend: "UCX"`

hchings replied:
Thanks for the catch. Updated.

pcastonguay commented on tests/integration/defs/llmapi/ray/test_llm_ray_e2e.py:
This file is empty, could we remove?

pcastonguay commented on tensorrt_llm/_torch/pyexecutor/executor_request_queue.py:
```diff
@@ -563,13 +571,26 @@ def _broadcast_new_requests(
 
         if not self.dist.is_last_pp_rank:
-            self.dist.send_object(payloads, self.dist.next_pp_rank, tag)
+            if self._disable_mpi:
+                works.extend(
+                    self.dist.isend_object(payloads, self.dist.next_pp_rank,
+                                           tag))
+            else:
+                self.dist.send_object(payloads, self.dist.next_pp_rank, tag)
+
+        for work in works:
+            try:
+                work.wait()
+            except Exception as e:
+                raise RuntimeError(
+                    f"Asynchronous broadcast operation failed: {e}") from e
```
Not sure I understand why we need the `works` list here, looks like only one item can be added. Could we refactor as:

```suggestion
            if self._disable_mpi:
                isend_payload = 
                    self.dist.isend_object(payloads, self.dist.next_pp_rank,
                                           tag)
                try:
                    isend_payload.wait()
                except Exception as e:
                    raise RuntimeError(
                        f"Asynchronous broadcast operation failed: {e}") from e                           
            else:
                self.dist.send_object(payloads, self.dist.next_pp_rank, tag)
```

hchings replied:
Updated. @pcastonguay PTAL, thanks.

hchings commented on tensorrt_llm/executor/ray_gpu_worker.py:
```diff
@@ -0,0 +1,540 @@
+        self.master_port = os.environ["MASTER_PORT"]
+
+        # Ray can't pickle TensorRT logger; import/use it inside methods only.
+        from tensorrt_llm.logger import logger
```
If we import it at the beginning of the file, Ray Actor will have pickle issue. @shuyixiong has changed it to global in [this commit](https://github.com/NVIDIA/TensorRT-LLM/pull/7520/commits/99d73a70f7fadf09eb8b4056b3b01205a981b5c2). PTAL.

hchings commented on tensorrt_llm/_torch/pyexecutor/py_executor.py:
```diff
@@ -941,7 +958,16 @@ def _executor_loop_pp(self):
     def wait_on_pp_send_handles(self, microbatch_id):
         if self.send_handles[microbatch_id] is not None:
-            self.send_handles[microbatch_id].wait()
+            # TODO: need clean up
```
Hi @pcastonguay, `self.send_handles[microbatch_id]` being a list in Ray path is because `TorchDist.isend_object` ([L355](https://github.com/NVIDIA/TensorRT-LLM/pull/7520/files#diff-af8032720a651060671a60f2f8a090a791e333ae88f59fb1c73db0512ddc574bR355)) has to send size and payload as two separate nonblocking ops with torch.distributed, which returns as a list of work handles.

Even if we change `TorchDist.isend_object()` to `torch.distributed.batch_isend_irecv`, the return type is still a list of handles ([doc](https://docs.pytorch.org/docs/main/distributed.html#torch.distributed.batch_isend_irecv)). 

One way we can do is to add a light wrapper on the return value of `isend_object()` that does the wait on a list of handles, so that we don't have to branch out for Ray v.s. MPI in `py_executor.py`.  

+ @tongyuantongyu  for other ideas.

hchings replied:
Re pp tests:

- In `H100_PCIe-PyTorch-Ray-1`, we'll reuse existing pp tests (e.g., [test_return_logits_pp](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tests/integration/defs/accuracy/test_llm_api_pytorch.py#L502)) for testing Ray orchestrator.  
     - I checked locally by running `pytest -k test_return_logits_pp[pp2-True] --run-ray` and it's currently passing.
- Added a simple pp2 test [here](https://github.com/NVIDIA/TensorRT-LLM/pull/7520/commits/51b40da2f8e8dac58f5fb2a2c84c95d9d5318ed9#diff-a541a49d024d69156249bf1b1143c6d5e4a68507320db0116ed0db5cb9d08cf9R18).

pcastonguay replied:
Yes I agree too.

shuyixiong commented on tensorrt_llm/_torch/pyexecutor/py_executor.py:
```diff
@@ -1797,6 +1823,14 @@ def _enqueue_responses(self, responses: Iterable[Tuple[int, LlmResponse]]):
+                            req_id] is not None:
+                        self.result_wait_queues[req_id].put_response.remote(
+                            resp.client_id, resp)
+                        self.result_wait_queues.pop(req_id, None)
```
Hi @pcastonguay, I have moved results_wait_queues cleanup to _do_terminate_request which should cover all normal cases and error cases.

hchings commented on examples/ray/llm_inference_async_ray.py:
```diff
@@ -0,0 +1,54 @@
+# Generate text asynchronously with Ray orchestrator.
```
@litaotju We can add this, but I don't see other llm-api examples have it though, so not sure what's the rule. Can you confirm or who should I check it with?

hchings commented on examples/ray/multi_nodes/README.md:
```diff
@@ -0,0 +1,41 @@
+    >> enroot exec <process id> bash
+
+    # Under your work directory:
+    >> pip install -e . # if needed
```
Yes. Updated the README wordings.

hchings commented on examples/ray/multi_nodes/run_cluster.sh:
```diff
@@ -0,0 +1,229 @@
+########################################################
+
+# enabled dashboard only for debug
+# Add cd tekit/ and pip install -e . to head_cmd and worker_cmd if needed
```
Updated. Thanks for catching this.

hchings commented on tensorrt_llm/_torch/distributed/communicator.py:
```diff
@@ -161,63 +169,382 @@ def pp_broadcast(self, obj, root=0):
+        pg_util_path = os.path.normpath(f"{src_path}/../../libs")
+        pg_broker = torch.utils.cpp_extension.load(
+            name="pg_broker",
+            sources=[f"{src_path}/pgBroker.cpp"],
```
Isn't this loading?  + @tongyuantongyu  to handle this one.

tongyuantongyu replied:
> Shall this pg broker extenstion be built during TRTLLM wheel build time, instead of on the fly?

See https://github.com/NVIDIA/TensorRT-LLM/pull/7520#discussion_r2354220271 for explanation. We also have another plan to avoid packaging this .cpp source,  with the cost of being slower.

> And what happens if multiple processing is building this one?

PyTorch should have safeguards to build it only once and safely.

hchings commented on tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py:
```diff
@@ -10,7 +10,7 @@
 
     @torch.library.register_fake("trtllm::allreduce")
-    def _(
+    def allreduce(
```
+ @tongyuantongyu @shuyixiong  - any context on this part?

tongyuantongyu replied:
We want to reuse the code in `allreduce_pg` by calling this function. So we need a unique name to reference it.

tongyuantongyu commented on tensorrt_llm/_torch/distributed/pgBroker.cpp:
```diff
@@ -0,0 +1,37 @@
+
+} // namespace tensorrt_llm::pg_broker
+
+PYBIND11_MODULE(TORCH_EXTENSION_NAME, m)
```
This is JIT-ed using pybind11 headers vendored by PyTorch.

shuyixiong commented on tensorrt_llm/_torch/distributed/communicator.py:
```diff
@@ -1,15 +1,22 @@
 from typing import Optional
 
 import numpy as np
+import ray
```
I have removed the dependency on ray in mpi path [here](https://github.com/joyang-nv/TensorRT-LLM/commit/976aa1388d83e15f0c39be45be35cfabb9a42a83) and tested it with a docker image that does not have ray installed.

hchings replied:
Preserving Shuyi's commit link here  https://github.com/NVIDIA/TensorRT-LLM/pull/7520/commits/976aa1388d83e15f0c39be45be35cfabb9a42a83  as I need to squash for rebase.

hchings commented on tensorrt_llm/llmapi/llm_args.py:
```diff
@@ -2335,6 +2335,13 @@ class TorchLlmArgs(BaseLlmArgs):
         status="prototype",
     )
 
+    orchestrator_type: Optional[str] = Field(
```
updated

hchings commented on tests/unittest/_torch/multi_gpu/test_allreduce.py:
```diff
@@ -574,3 +579,120 @@ def test_moe_finalize_allreduce_patterns(mpi_pool_executor):
+
+
+@ray.remote(num_gpus=1)
+class AllreducePGTest:
```
Agree. @shuyixiong can help moving all ray multi gpu tests into a single `test_ops_ray.py`.

HuiGao-NV reviewed (approved):
LGTM

hchings commented on tests/unittest/_torch/multi_gpu/test_allgather.py:
```diff
@@ -0,0 +1,130 @@
+import os
+
```
Shuyi has moved this. Please see https://github.com/NVIDIA/TensorRT-LLM/pull/7520/commits/cd1f17627a2d6c5a444b942243047d80d936a7ae#diff-47c132b0599d6a31a397f3477d25990e587d747a82af0622fb136120602e8330

tongyuantongyu committed:
feat: Ray integration

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

Add ray cleanup fixture

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

Add Ray test cases

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

address comments and fixes

Signed-off-by: Erin Ho <14718778+hchings@users.noreply.github.com>

add tests for ray examples, refactor to BaseWorker

Signed-off-by: Erin Ho <14718778+hchings@users.noreply.github.com>

Add ray example tests to CI list

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

Fix CUDA Graph + PG NCCL Coalescing

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

Unify WorkerExit

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

review cleanup

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

Make sub_pg inherit backend from global pg and some cleanup

Signed-off-by: shuyix <219646547+shuyixiong@users.noreply.github.com>

Cleanup

Signed-off-by: shuyix <219646547+shuyixiong@users.noreply.github.com>

CI fix

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

fix single gpu and api stability tests

Signed-off-by: Erin Ho <14718778+hchings@users.noreply.github.com>

Fix fake ray import

Signed-off-by: shuyix <219646547+shuyixiong@users.noreply.github.com>

Fix device mesh on single gpu

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

build & CI review

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

Move ray requirements declaration

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

Fix error response handling

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

Remove failing case for now

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

skip get stats tests

Signed-off-by: Erin Ho <14718778+hchings@users.noreply.github.com>

skip test_fp8_block_scales_4gpus

Signed-off-by: Erin Ho <14718778+hchings@users.noreply.github.com>

Resolve rebase conflict and revert result_wait_queue cleanup

Signed-off-by: shuyix <219646547+shuyixiong@users.noreply.github.com>

fix test_disaggregated_ctx**_gen**

Signed-off-by: Erin Ho <14718778+hchings@users.noreply.github.com>

Remove empty case

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

fix ci tests

Signed-off-by: Erin Ho <14718778+hchings@users.noreply.github.com>

Back result_wait_queue cleanup

Signed-off-by: shuyix <219646547+shuyixiong@users.noreply.github.com>

Update jenkins/L0_Test.groovy

Co-authored-by: Yanchao Lu <yanchaol@nvidia.com>
Signed-off-by: shuyixiong <219646547+shuyixiong@users.noreply.github.com>

Move ray stage using 4 gpus to b200

Signed-off-by: shuyix <219646547+shuyixiong@users.noreply.github.com>

Add orchestrator mpi to other test groups in b200 yaml

Signed-off-by: shuyix <219646547+shuyixiong@users.noreply.github.com>

Skip get perf metrics tests

Signed-off-by: shuyix <219646547+shuyixiong@users.noreply.github.com>

Skip unsupported tests in ray stage

Signed-off-by: shuyix <219646547+shuyixiong@users.noreply.github.com>

minor test fix & nit

Signed-off-by: Erin Ho <14718778+hchings@users.noreply.github.com>

skip newly added test

Signed-off-by: Erin Ho <14718778+hchings@users.noreply.github.com>

skip test_disaggregated_serving.py in ray stage until we add cleanup

Superjomn commented on tensorrt_llm/executor/result.py:
```diff
@@ -146,12 +152,104 @@ def logprobs_diff(self) -> TokenLogprobs | List[float]:
     def __init__(self,
                  id: int,
                  sampling_params: SamplingParams,
+                 queue: Optional[RayAsyncQueue] = None,
```
Is this argument specific to Ray? To make it clearer, maybe we can either
- Renaming it to `ray_queue`
- Or, expose both `queue` and `aqueue` arguments for general purpose, and leave creating the aqueue and queue instances for different scenarios to the upper level.

hchings replied:
It's specific to Ray. Changing to `ray_queue`.

Superjomn commented on tensorrt_llm/executor/result.py:
```diff
@@ -381,6 +489,9 @@ def _handle_response(self,
         else:
             raise ValueError(f"Unknown response type: {response}")
 
+        if self._done and mpi_disabled():
```
How about protecting this with a Ray flag or an assert? In mpi path, the queue doesn't have an `unregister` method, which makes it confusing if this if-statement is not protected well.
Maybe the following is better:
```python
assert hasattr(self.queue, "unregister"), "Ray path should be activated"
```

hchings commented on cpp/tensorrt_llm/executor/cache_transmission/ucx_utils/ucxCacheCommunicator.h:
```diff
@@ -78,6 +80,11 @@ class UcxConnectionManager : public ConnectionManager, public std::enable_shared
     std::vector<Connection const*> getConnections(CommState const& state) override;
     [[nodiscard]] CommState const& getCommState() const override;
+
+    [[nodiscard]] int getRank() const
+    {
+        return mRank;
+    }
```
not doing

laikhtewari reviewed (approved):
to_delete dir should be removed asap

tensorrt-cicd commented:
[PR_Github #19386](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19386/) [ run ] triggered by Bot

tongyuantongyu commented:
/bot run --stage-list "H100_PCIe-PyTorch-Ray-1, DGX_H100-2_GPUs-PyTorch-Ray-1, DGX_H100-4_GPUs-PyTorch-Ray-1" --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #19391](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19391/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19392](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19392/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19392](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19392/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #14567 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14567/) completed with status: 'FAILURE'

tongyuantongyu commented:
/bot run --stage-list "DGX_H100-2_GPUs-PyTorch-Ray-1, DGX_H100-4_GPUs-PyTorch-Ray-1" --disable-fail-fast --only-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #19394](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19394/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #14569 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14569/) completed with status: 'FAILURE'

tongyuantongyu commented:
/bot run --stage-list "DGX_H100-4_GPUs-PyTorch-Ray-1" --disable-fail-fast --only-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #19507](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19507/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19507](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19507/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #14662 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14662/) completed with status: 'FAILURE'

Superjomn reviewed (approved):
LGTM on llmapi changes

tongyuantongyu commented on cpp/include/tensorrt_llm/runtime/utils/pgUtils.h:
```diff
@@ -0,0 +1,287 @@
+    return at::from_blob(data, {static_cast<int64_t>(size)}, c10::TensorOptions{}.dtype(torch::kChar));
+}
+
+template <typename T>
+torch::Tensor wrap_tensor(T const* data, size_t size)
+{
+    return wrap_tensor(const_cast<T*>(data), size);
+}
+
```
It's an unfortunate fact that `at::from_blob` does not accept const pointer. The only option is to cast away the const, create a Tensor and carefully not to write to it.

tensorrt-cicd commented:
[PR_Github #19784](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19784/) [ run ] triggered by Bot

litaotju commented on tests/integration/test_lists/test-db/l0_h100.yml:
```diff
@@ -189,6 +216,7 @@ l0_h100:
     terms:
       stage: post_merge
       backend: pytorch
+      orchestrator: mpi
```
A nit. 
Isn't the default using mpi? So is this really needed?

tongyuantongyu replied:
Mako's behavior, which I assume is for backend compatibility, is that unspecified tests will always be matched: https://github.com/NVIDIA/TensorRT-LLM/blob/60101eb8a56e3671f3f2f72217588f81e2facda2/jenkins/L0_Test.groovy#L1252.

That means if we don't mark these with `orchestrator: mpi`, they will run in Ray stages as well.

tongyuantongyu commented:
/bot run --disable-fail-fast

chzblych commented on tests/integration/test_lists/test-db/l0_dgx_h100.yml:
```diff
@@ -171,6 +175,7 @@ l0_dgx_h100:
       stage: pre_merge
       backend: cpp
       auto_trigger: others
+      orchestrator: mpi
```
It's better to be consistent. E.g., adding `orchestrator: mpi` or `orchestrator: ray` everywhere in the test-db file, or only adding either of them when it's necessary.

This comment is the same for `l0_dgx_h100.yml` and `l0_h100.yml`.

tensorrt-cicd commented:
[PR_Github #19935](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19935/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15008](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15008/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #19975](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19975/) [ run ] triggered by Bot

hchings commented on jenkins/L0_Test.groovy:
```diff
@@ -1708,6 +1714,11 @@ def runLLMTestlistOnPlatformImpl(pipeline, platform, testList, config=VANILLA_CO
                 "--perf-log-formats yaml"
             ]
         }
+        if (stageName.contains("-Ray-")) {
```
IMO having separate stages for Ray is cleaner, given right now Ray is at prototype stage and an optional path that still need to collect customer feedbacks.

@chzblych - Is the overhead of having Ray tests in separate stages significant?

tongyuantongyu replied:
We want to be able to add any test to list to test it with Ray. If using fixture, for EACH test we want to run, we need to update the case definition and ALL occurrences of the testcase name in various test lists.

Any idea we can do it without updating the test definition?

chzblych replied:
It's fair. I'm fine to keep the `Ray` test stages.

hchings commented on jenkins/L0_Test.groovy:
```diff
@@ -1708,6 +1714,11 @@ def runLLMTestlistOnPlatformImpl(pipeline, platform, testList, config=VANILLA_CO
+        if (stageName.contains("-Ray-")) {
+            testCmdLine += ["--run-ray"]
+
+            sh "pip3 install ray[default]"
```
This is due to previous alignments w/ other reviewers that we don't want to have ray installed nor imported when users are using the default MPI path. 

In the [README](https://github.com/NVIDIA/TensorRT-LLM/pull/7520/files/834842f0cfb0ed198edb4470e04385b6085c423f#diff-e7412e3deb5417508efd1d240953d40edb98a4b3b61abc677975a5a584c647b7R17) under `examples/ray_orchestrator/`, it has Ray installation instruction, which IMO is sufficient at this stage.

tongyuantongyu replied:
Ray is a quite heavy dependency that only users would like to try it need it installed:
```
Downloading ray-2.49.2-cp310-cp310-manylinux2014_x86_64.whl (69.9 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 69.9/69.9 MB 76.7 MB/s eta 0:00:00
```

So we decided to make it optional: https://github.com/NVIDIA/TensorRT-LLM/pull/7520#discussion_r2353912113.

Let me add it as optional extra so user can use `pip install tensorrt_llm[ray]`.

hchings replied:
Discussed off line, we'll add a requirements.txt in `examples/ray_orchestrator/` and only set up `pip install tensorrt_llm[ray]` once Ray path is more mature perf wise.

chzblych replied:
Thanks for the explanation! The approach looks good to me.

hchings commented on jenkins/L0_Test.groovy:
```diff
@@ -2043,6 +2054,8 @@ def launchTestJobs(pipeline, testFilter)
         "DGX_H100-2_GPUs-PyTorch-Others-1": ["dgx-h100-x2", "l0_dgx_h100", 1, 1, 2],
         "DGX_H100-4_GPUs-PyTorch-GptOss-1": ["dgx-h100-x4", "l0_dgx_h100", 1, 1, 4],
         "DGX_H100-4_GPUs-PyTorch-Others-1": ["dgx-h100-x4", "l0_dgx_h100", 1, 1, 4],
+        "DGX_H100-2_GPUs-PyTorch-Ray-1": ["dgx-h100-x2", "l0_dgx_h100", 1, 1, 2],
```
@tongyuantongyu suggested that we move Ray multi-GPUs to post-merge instead. @chzblych lmk if you still have concern on this.

tongyuantongyu replied:
Is it OK to run them in post-merge? We can do it after the pipeline is clean in this PR.

chzblych replied:
Starting from the post-merge pipeline sounds good to me.

tongyuantongyu commented:
/bot run --disable-fail-fast --only-multi-gpu-test --stage-list "DGX_H100-4_GPUs-PyTorch-Ray-1"

tensorrt-cicd commented:
[PR_Github #20080](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20080/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15128 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15128/) completed with status: 'FAILURE'

hchings commented:
/bot run --disable-fail-fast --only-multi-gpu-test --stage-list "DGX_H100-4_GPUs-PyTorch-Ray-1"

tensorrt-cicd commented:
[PR_Github #20261](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20261/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20261](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20261/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15278 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15278/) completed with status: 'FAILURE'

hchings commented:
/bot run --disable-fail-fast --only-multi-gpu-test --stage-list "DGX_H100-4_GPUs-PyTorch-Ray-1"

shuyixiong commented:
Local perf testing confirms this PR has no negative impact on MPI perf. Detailed results are attached below.[perf_data_2.csv](https://github.com/user-attachments/files/22609543/perf_data_2.csv)

tongyuantongyu commented on tests/integration/test_lists/test-db/l0_dgx_h100.yml:
```diff
@@ -177,6 +181,7 @@ l0_dgx_h100:
       stage: pre_merge
       backend: cpp
       auto_trigger: others
+      orchestrator: mpi
```
This line excludes these test cases from the Ray stage. We only want to run a few selected tests that are representative.

shuyixiong commented:
/bot run

tensorrt-cicd commented:
[PR_Github #20323](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20323/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20323](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20323/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15328](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15328/) completed with status: 'FAILURE'

hchings commented:
/bot run

joyang-nv commented:
/bot run

tensorrt-cicd commented:
[PR_Github #20454](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20454/) [ run ] completed with state `ABORTED`
[/LLM/main/L0_MergeRequest_PR pipeline #15420](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15420/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #20511](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20511/) [ run ] triggered by Bot

hchings commented:
/bot kill

tensorrt-cicd commented:
[PR_Github #20513](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20513/) [ kill ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20513](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20513/) [ kill ] completed with state `SUCCESS`
Successfully killed previous jobs for commit `40026de`

tensorrt-cicd commented:
[PR_Github #20524](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20524/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20530](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20530/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20531](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20531/) [ kill ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20530](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20530/) [ run ] completed with state `ABORTED`
<a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15488/" target="_blank">LLM/main/L0_MergeRequest_PR #15488</a> <a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15488//display/redirect" target="_blank">(Blue Ocean)</a> completed with status: ABORTED

tensorrt-cicd commented:
[PR_Github #20556](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20556/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20556](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20556/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15513](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15513/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #20563](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20563/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20567](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20567/) [ kill ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20563](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20563/) [ run ] completed with state `ABORTED`
<a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15520/" target="_blank">LLM/main/L0_MergeRequest_PR #15520</a> <a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15520//display/redirect" target="_blank">(Blue Ocean)</a> completed with status: ABORTED

hchings committed:
force local cluster

Signed-off-by: Erin Ho <14718778+hchings@users.noreply.github.com>

hchings commented:
/bot run --stage-list "H100_PCIe-PyTorch-Ray-1, DGX_H100-2_GPUs-PyTorch-Ray-1, DGX_H100-4_GPUs-PyTorch-Ray-1"

tensorrt-cicd commented:
[PR_Github #20573](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20573/) [ run ] triggered by Bot

joyang-nv commented:
/bot help

joyang-nv commented:
/bot run --stage-list "DGX_B200-4_GPUs-PyTorch-Ray-1,DGX_H100-2_GPUs-PyTorch-Ray-1,H100_PCIe-PyTorch-Ray-1"

tensorrt-cicd commented:
[PR_Github #20574](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20574/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20574](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20574/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15529 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15529/) completed with status: 'SUCCESS'

tensorrt-cicd commented:
[PR_Github #20579](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20579/) [ kill ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20579](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20579/) [ kill ] completed with state `SUCCESS`
Successfully killed previous jobs for commit `efc9b16`

tensorrt-cicd commented:
[PR_Github #20580](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20580/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20587](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20587/) [ kill ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20587](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20587/) [ kill ] completed with state `SUCCESS`
Successfully killed previous jobs for commit `efc9b16`

tensorrt-cicd commented:
[PR_Github #20588](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20588/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20608](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20608/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15557](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15557/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/15557/test-results/rerun_report.html) for details.

joyang-nv committed:
Merge branch 'main' into user/joyang/ray

joyang-nv commented on tensorrt_llm/_torch/distributed/pg_utils.py:
```diff
@@ -0,0 +1,41 @@
+    # Create subgroup under the provided pg; ranks are global ranks
+    sub_pg = dist.new_group(ranks=ranks,
+                            use_local_synchronization=True,
+                            backend='gloo')
```
No problem at now.
